### PR TITLE
[Part3] grpc: add grpc behavior

### PR DESCRIPTION
**Pull request by vivekrsharma**
Adds grpc behavior support over proto and json encodings.

### CODE REVIEW

File: yarpc-crossdock/src/main/java/io/yarpc/crossdock/client/grpc/EchoGrpcBehavior.java [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+        String wantToken = "abc";
+        String gotToken = client.echo(wantToken);
+        if (wantToken.equals(gotToken)) {
+          return CompletableFuture.completedFuture(new CrossdockResponse().success(wantToken));
+        }
+        return CompletableFuture.completedFuture(
+            new CrossdockResponse().error("Expected " + wantToken + " but got " + gotToken));
+
+      default:
+        return CompletableFuture.completedFuture(null);
```
Comment (bpon) [inline yarpc-crossdock/src/main/java/io/yarpc/crossdock/client/grpc/EchoGrpcBehavior.java]: Shouldn't this be `new CrossdockResponse().skipped(...)`?

File: yarpc-crossdock/src/main/java/io/yarpc/crossdock/client/grpc/EchoGrpcBehavior.java [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+public class EchoGrpcBehavior implements Behavior {
+
+  @SuppressFBWarnings("NP_NONNULL_PARAM_VIOLATION")
+  @Override
+  public CompletionStage<CrossdockResponse> run(CrossdockRequest request) {
+
+    switch (request.getParam("encoding")) {
+      case "raw":
+        return CompletableFuture.completedFuture(
+            new CrossdockResponse().skipped("raw behavior not supported"));
```
Comment (bpon) [inline yarpc-crossdock/src/main/java/io/yarpc/crossdock/client/grpc/EchoGrpcBehavior.java]: I feel like this should be the `default:` case.  See below.

Comment (vivekrsharma) [inline yarpc-crossdock/src/main/java/io/yarpc/crossdock/client/grpc/EchoGrpcBehavior.java]: Makes sense, raw and thrift in following PRs and can add these cases then.

File: yarpc-crossdock/src/main/java/io/yarpc/crossdock/client/grpc/EchoGrpcBehavior.java [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+import java.util.concurrent.TimeUnit;
+import lombok.extern.slf4j.Slf4j;
+import works.crossdock.CrossdockRequest;
+import works.crossdock.CrossdockResponse;
+import works.crossdock.client.Behavior;
+
+@Slf4j
+public class EchoGrpcBehavior implements Behavior {
+
+  @SuppressFBWarnings("NP_NONNULL_PARAM_VIOLATION")
```
Comment (bpon) [inline yarpc-crossdock/src/main/java/io/yarpc/crossdock/client/grpc/EchoGrpcBehavior.java]: You don't need this anymore.

File: yarpc-crossdock/src/main/java/io/yarpc/crossdock/client/grpc/EchoGrpcBehavior.java [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+        String gotToken = client.echo(wantToken);
+        if (wantToken.equals(gotToken)) {
+          return CompletableFuture.completedFuture(new CrossdockResponse().success(wantToken));
+        }
+        return CompletableFuture.completedFuture(
+            new CrossdockResponse().error("Expected " + wantToken + " but got " + gotToken));
+
+      default:
+        return CompletableFuture.completedFuture(
+            new CrossdockResponse().skipped(encoding + "behavior not supported"));
```
Comment (bpon) [inline yarpc-crossdock/src/main/java/io/yarpc/crossdock/client/grpc/EchoGrpcBehavior.java]: Missing a space before "behavior".

File: yarpc-crossdock/src/main/java/io/yarpc/crossdock/client/grpc/EchoGrpcBehavior.java [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+    switch (encoding) {
+      case "json":
+        ManagedChannel channel =
+            ManagedChannelBuilder.forAddress(request.getParam("server"), 8089)
+                .usePlaintext(true)
+                .build();
+        Ping ping = Ping.newBuilder().setBeep("abc").build();
+
+        Metadata metadata = GrpcUtils.setupYarpcHeaders();
+        metadata.put(Metadata.Key.of("Rpc-Encoding", Metadata.ASCII_STRING_MARSHALLER), "json");
```
Comment (peter-edge) [inline yarpc-crossdock/src/main/java/io/yarpc/crossdock/client/grpc/EchoGrpcBehavior.java]: You should make `Rpc-Encoding` etc `public static final` somewhere. Also they all should be lowercase.

File: yarpc-crossdock/src/main/java/io/yarpc/crossdock/client/grpc/GrpcUtils.java [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+  /**
+   * Setups the custom yarpc headers required by all the callers. Callers are still required to set
+   * the encoding used.
+   *
+   * @return Metadata object that contains the custom yarpc headers.
+   */
+  public static Metadata setupYarpcHeaders() {
+    Metadata metadata = new Metadata();
+    metadata.put(Metadata.Key.of("Rpc-Caller", Metadata.ASCII_STRING_MARSHALLER), "client");
```
Comment (peter-edge) [inline yarpc-crossdock/src/main/java/io/yarpc/crossdock/client/grpc/GrpcUtils.java]:
```java
public class GrpcUtils

  public static final String rpcCallerHeaderKey = "rpc-caller";
  ...
}
```

File: yarpc-crossdock/src/main/java/io/yarpc/crossdock/client/grpc/EchoGrpcBehavior.java (hunk 5 above)
Comment (vivekrsharma) [inline yarpc-crossdock/src/main/java/io/yarpc/crossdock/client/grpc/EchoGrpcBehavior.java]: Done.

File: yarpc-crossdock/src/main/java/io/yarpc/crossdock/client/grpc/GrpcUtils.java (hunk 6 above)
Comment (vivekrsharma) [inline yarpc-crossdock/src/main/java/io/yarpc/crossdock/client/grpc/GrpcUtils.java]: SGTM.

File: yarpc-crossdock/src/main/java/io/yarpc/crossdock/client/grpc/EchoGrpcBehavior.java [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+import works.crossdock.client.Behavior;
+
+@Slf4j
+public class EchoGrpcBehavior implements Behavior {
+  public static final String JSON_ENCODING = "json";
+
+  @Override
+  public CompletionStage<CrossdockResponse> run(CrossdockRequest request) {
+    String encoding = request.getParam("encoding");
+    switch (encoding) {
```
Comment (peter-edge) [inline yarpc-crossdock/src/main/java/io/yarpc/crossdock/client/grpc/EchoGrpcBehavior.java]: You have a constant for this

Comment (vivekrsharma) [inline yarpc-crossdock/src/main/java/io/yarpc/crossdock/client/grpc/EchoGrpcBehavior.java]: isn't that "rpc-encoding" and not "encoding"

**Pull request merged**